massgen/frontend/displays/tui_event_pipeline.py

```
@@ -289,18 +311,35 @@ def _apply_tool_output(self, output: ContentOutput, round_number: int, timeline:
                 except Exception:
                     existing = None
 
-                if existing is None:
+                # Avoid duplicating tools that already live inside a batch
+                in_batch = output.batch_action == "update_batch"
+                if not in_batch:
+                    try:
+                        in_batch = timeline.get_tool_batch(tool_data.tool_id) is not None
+                    except Exception:
```

Ruff (0.14.14)

[warning] Do not catch blind exception: `Exception`

(BLE001)